collator/src/lib.rs

```diff
@@ -215,197 +210,204 @@ where
-			let proposer = proposer_future
-				.await
-				.map_err(|e| {
-					error!(
-						target: "cumulus-collator",
-						"Could not create proposer: {:?}",
-						e,
-					)
-				})
-				.ok()?;
-
-			let inherent_data = Self::inherent_data(
-				inherent_providers,
-				global_validation,
-				local_validation,
-				downward_messages,
-			)?;
-
-			let Proposal {
-				block,
-				storage_changes,
-				proof,
-			} = proposer
-				.propose(
-					inherent_data,
-					Default::default(),
-					//TODO: Fix this.
-					Duration::from_millis(500),
-					RecordProof::Yes,
-				)
-				.await
-				.map_err(|e| {
-					error!(
-						target: "cumulus-collator",
-						"Proposing failed: {:?}",
-						e,
-					)
-				})
-				.ok()?;
-
-			let proof = match proof {
-				Some(proof) => proof,
-				None => {
-					error!(
-						target: "cumulus-collator",
-						"Proposer did not return the requested proof.",
-					);
+			let new_validation_code = sp_io::storage::get(well_known_keys::NEW_VALIDATION_CODE);
 
+			Some(Collation {
+				//TODO: What should we do here?
+				fees: 0,
+				upward_messages,
+				new_validation_code: new_validation_code.map(Into::into),
+				head_data,
+				proof_of_validity: PoV { block_data },
+			})
+		})
+	}
+
+	async fn produce_candidate(
+		mut self,
+		_: PHash,
+		validation_data: ValidationData,
+	) -> Option<Collation> {
+		trace!(target: "cumulus-collator", "Producing candidate");
+
+		let last_head =
+			match Block::Header::decode(&mut &validation_data.persisted.parent_head.0[..]) {
+				Ok(x) => x,
+				Err(e) => {
+					error!(target: "cumulus-collator", "Could not decode the head data: {:?}", e);
 					return None;
 				}
 			};
 
-			let (header, extrinsics) = block.deconstruct();
+		if !self.check_block_status(last_head.hash()) {
+			return None;
+		}
 
-			// Create the parachain block data for the validators.
-			let b = ParachainBlockData::<Block>::new(header.clone(), extrinsics, proof);
+		let proposer_future = self.proposer_factory.lock().init(&last_head);
 
-			let mut block_import_params = BlockImportParams::new(BlockOrigin::Own, header);
-			block_import_params.body = Some(b.extrinsics().to_vec());
-			// Best block is determined by the relay chain.
-			block_import_params.fork_choice = Some(ForkChoiceStrategy::Custom(false));
-			block_import_params.storage_changes = Some(storage_changes);
+		let proposer = proposer_future
+			.await
+			.map_err(|e| {
+				error!(
+					target: "cumulus-collator",
+					"Could not create proposer: {:?}",
+					e,
+				)
+			})
+			.ok()?;
 
-			if let Err(err) = block_import
-				.lock()
-				.import_block(block_import_params, Default::default())
-			{
+		let inherent_data = self.inherent_data(
+			&validation_data,
+			// TODO get the downward messages
+			Vec::new(),
+		)?;
+
+		let Proposal {
+			block,
+			storage_changes,
+			proof,
+		} = proposer
+			.propose(
+				inherent_data,
+				Default::default(),
+				//TODO: Fix this.
+				Duration::from_millis(500),
```

Comment:
what would this fix look like?

Reply:
This comment is old. However, I think that I will be able to improve this with the switch to the slot interface. In the end I just need a reasonable max time for the block production. However, we will probably stick to 500ms for block production? This gives us 1.5 seconds for propagating it to the parachain validators and validation on their end.